src/server/game-server.ts

```diff
@@ -20,15 +59,31 @@ export class GameServer {
   // The unprocessed inputs for each player
   frameBuffers: {[id:string]: InputFrame[]} = {};
 
-  constructor(sockets: SocketIO.Socket[]) {
+  constructor(gameFinishedCallback: ()=>void) {
+    this.gameFinishedCallback = gameFinishedCallback;
     this.game = Game.init();
     this.lastTick = Date.now();
-    this.sockets = sockets;
+  }
+
+  addPlayer(socket: SocketIO.Socket) {
+    this.sockets.push(socket);
+    this.connectedPlayers += 1;
+    socket.on('disconnect', () => this.onPregameDisconnection(socket));
+    if (this.sockets.length == Game.settings.maxPlayers) {
+      this.startGame();
+    }
+    for (let s of this.sockets) {
+      s.emit(NEW_PLAYER_JOINED, this.sockets.length);
+    }
+  }
+
+  startGame() {
     for (var socket of this.sockets) {
       let player = Game.addPlayer(this.game);
       socket.on(SEND_FRAMES, (frames: InputFrame[]) => this.acceptFrames(frames, player.id));
+      socket.removeAllListeners('disconnect');
```

Comment:
Might be better to have
`onDisconnection` remove them from `this.sockets`
And then have an `onInGameDisconnection` that has any extra logic needed 
Then you wouldn't have to clear any listeners, and you'd have less code duplication.